src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp

```diff
@@ -90,13 +90,24 @@ void QuaternionFunctionOfTime<MaxDeriv>::solve_quaternion_ode(
     const gsl::not_null<boost::math::quaternion<double>*>
         quaternion_to_integrate,
     const double t0, const double t) const {
+  // Boost is stupid and assumes the times it's integrating between are order
+  // unity. So if t or t0 is > 10, then you can't take t-t0 to machine precision
+  // which causes roundoff error within boost. It was found experimentally that
+  // an infinite loop *within boost* is possible if such a condition is met.
+  // Thus, we rescale the times and the RHS by this factor to ensure the times
+  // are of order unity to avoid this infinite loop. The resulting quaternion is
+  // left unchanged.
+  const double factor = std::max(1.0, std::max(t0, t));
```

Comment:
Optional: `std::max({1.0, t0, t})`

I assume because this is control system stuff we don't have to worry about either of those being negative.

Reply:
Do the initializer-lists do a memory allocation in a case like this?

Reply:
From [dcl.init.list]: "An object of type `std::initializer_list<E>` is constructed from an initializer list as if the implementation generated and materialized a prvalue of type 'array of N `const E`'...and the `std::initializer_list<E>` object is constructed to refer to that array."  From [initializer_list.syn]: "An object of type `initializer_list<E>` provides access to an array of objects of type `const E` [Note: A pair of pointers or a pointer plus a length would be obvious representations for `initializer_list`...]"

So `std::initializer_list` is basically a view into an array (with the usual "as-if" rule caveat), so it never allocates memory itself (although its contents are copied and become const, which will have whatever side-effects that has normally).

Reply:
Thanks!

Reply:
I think I'll leave it as is. And @wthrowe correct, control systems should only be used in forward-in-time evolutions, so they will be positive.